Deduplication PR

hexylena opened the pull request:
Original PR message:

> I ran a code duplication tool over `lib/` and found some pretty awful copy-pasta. This PR cleans that up a little. It's probably a sign of a bigger architectural issue and this is just a bandaid.
> 
> NB: I have not manually tested the code to verify it still functions correctly, so feel free to decline this PR on those grounds, if no one else is interested in testing.

I'm expanding this to include other deduplication changes. They should probably get some more testing. If you're reviewing this, I strongly recommend viewing it commit-by-commit to understand where things are getting refactored from/to

hexylena committed:
Dedupe one of the worse offenders

hexylena committed:
Refactor sleeper duplications

hexylena committed:
Refactor identical role queries into a single function

hexylena committed:
Refactor copy-pasted function into separate module

hexylena committed:
Merge branch 'big-dedup' into deduplication

martenson commented on lib/galaxy/webapps/tool_shed/controllers/repository.py:
```diff
@@ -157,21 +174,10 @@ def browse_deprecated_repositories_i_own( self, trans, **kwd ):
-                return trans.response.send_redirect( web.url_for( controller='repository',
-                                                                  action='deprecate',
-                                                                  **kwd ) )
+        _redir = self._redirect_if_necesseary( trans, **kwd )
```
copy pasta typo (:

hexylena replied:
heh, indeed. Nice catch.

hexylena committed:
Refactor duplicate calculation code in reports

hexylena committed:
Correct copy-pasted typo

hexylena committed:
Make flake8 happy

hexylena committed:
Refactor into a single method and flake

hexylena committed:
refactor completely duplicated numpreds/successor processing

hexylena committed:
Fix formatting error

hexylena commented:
Okay, that's the bulk of the _major_ duplications (other than repository_grids.py, but that's not worth the time investment). 
- pulsar/lwr are highly duplicated
- Other things require extensive work to refactor/are architectural issues

Removing WIP tag as that's all the time I have for this right now.

nsoranzo commented on lib/galaxy/security/__init__.py:
```diff
@@ -208,11 +214,7 @@ def get_all_roles( self, trans, cntrller ):
-                                                       self.model.Role.table.c.type != self.model.Role.types.PRIVATE,
-                                                       self.model.Role.table.c.type != self.model.Role.types.SHARING ) ) \
-                                        .order_by( self.model.Role.table.c.name ):
+            for role in self._get_roles( trans):
```
s/_get_roles/_get_npns_roles/

nsoranzo commented on lib/galaxy/webapps/util.py:
```diff
@@ -0,0 +1,17 @@
+    """
+    formatters = []
+    # Formatter for mako
+    import mako.exceptions
```
Can you please move this at the top of the file, preceded by:

``` python
from galaxy import eggs
eggs.require("Mako")
```

hexylena committed:
Correct typo'd _get_roles function

hexylena committed:
Correct import for @nsoranzo

nsoranzo committed:
Fix wrong local import of galaxy.

hexylena committed:
Merge pull request #8 from nsoranzo/fix_for_erasche

Fix wrong local import of galaxy.

nsoranzo commented on lib/tool_shed/scripts/bootstrap_tool_shed/create_user_with_api_key.py:
```diff
@@ -65,27 +67,10 @@ def create_api_key( app, user ):
-    user_info_config = os.path.abspath( os.path.join( app.config.root, 'lib/tool_shed/scripts/bootstrap_tool_shed', 'user_info.xml' ) )
-    email = None
-    password = None
-    username = None
```
I think these 3 initialization lines should be added to `.bootstrap_util.admin_user_info()`.

hexylena replied:
Okay. Can do. There were two equivalent implementations so I picked the cleanest one. I can see a case where it wouldn't set those variables, so I'll copy the lines.

hexylena committed:
Initialized variables

hexylena committed:
Merge branch 'deduplication' of github.com:erasche/galaxy into deduplication

nsoranzo commented:
:+1: 

Thanks @erasche!

"194 additions and 450 deletions" = Cool!

nsoranzo committed:
Fix Tool Shed test broken with PR #706.

nsoranzo committed:
Fix MissingToolException when running "./run_tests.sh -api" without "-with_framework_test_tools".

nsoranzo committed:
Specify the tool_id of the missing tool when raising MissingToolException.

nsoranzo committed:
Fix "KeyError: name" error when running "./run_tests.sh -api".

Example traceback for test/api/test_workflows.py:WorkflowsApiTestCase.test_workflow_run_dynamic_output_collections :

galaxy.web.framework.decorators: ERROR: Uncaught exception in exposed API method:
Traceback (most recent call last):
  File "/opt/galaxy/lib/galaxy/web/framework/decorators.py", line 260, in decorator
    rval = func( self, trans, *args, **kwargs)
  File "/opt/galaxy/lib/galaxy/webapps/galaxy/api/workflows.py", line 182, in create
    return self.__api_import_new_workflow( trans, payload, **kwd )
  File "/opt/galaxy/lib/galaxy/webapps/galaxy/api/workflows.py", line 374, in __api_import_new_workflow
    workflow, missing_tool_tups = self._workflow_from_dict( trans, data, **from_dict_kwds )
  File "/opt/galaxy/lib/galaxy/web/base/controller.py", line 1111, in _workflow_from_dict
    publish=publish
  File "/opt/galaxy/lib/galaxy/managers/workflows.py", line 165, in build_workflow_from_dict
    missing_tool_tup = ( step_dict[ 'tool_id' ], step_dict[ 'name' ], step_dict[ 'tool_version' ] )
KeyError: 'name'

nsoranzo committed:
Print to stderr when calling sys.exit(1)

nsoranzo committed:
Use __file__ when adding paths to sys.path instead of os.getcwd() .

Also insert additional paths in position 1 unless strictly necessary.

bgruening committed:
Merge pull request #720 from nsoranzo/misc_fixes

Misc fixes

bgruening committed:
Merge pull request #722 from dannon/style_fix

Fix lint failure in toolshed/admin.

jmchilton committed:
Merge remote-tracking branch 'erasche/deduplication' into deduplication-merge

Conflicts:
	lib/tool_shed/scripts/bootstrap_tool_shed/bootstrap_util.py
	lib/tool_shed/scripts/bootstrap_tool_shed/create_user_with_api_key.py

jmchilton committed:
Fix lint problems from merge.

How did that happen, those pieces of code were largely unrelated. Ekkk...

jmchilton commented:
Pushed a conflict resolution merge for this. :+1: on this, will merge later today unless someone complains. Also thanks Eric!!

jmchilton commented:
About the Pulsar/LWR duplication - that is largely intentional. The LWR code in Galaxy will sit unmodified until finally removed. Modifying any of it does not make sense to me, it might break something and it is all deprecated. It might be time to remove the LWR code altogether just to clean up stuff though, I'll post to -dev about it.

hexylena commented:
@jmchilton thanks for the conflict-resolution merge! 

Yep, figured that it was intentional and thanks for posting to dev. Would be nice to get rid of legacy cruft :)
